### ISSUE TITLE
I'm not sure if this is expected behavior,  no deadlock is detected

### ISSUE THREAD

**Issue by HuMtk** [rust-lang/rust#144105]
I tried this code:

```rust
use std::sync::Mutex;

fn main() {
    let x = Mutex::new(0);
    println!("{}, {} ", x.lock().unwrap(), x.lock().unwrap());
}

```

I expected to see this happen: *I thought this should have deadlock detection*

Instead, this happened: *But he did not detect deadlock*

### Meta

`rustc --version --verbose`:
```
rustc 1.87.0 (17067e9ac 2025-05-09)
binary: rustc
commit-hash: 17067e9ac6d7ecb70e50f92c1944e545188d2359
commit-date: 2025-05-09
host: x86_64-pc-windows-gnu
release: 1.87.0
LLVM version: 20.1.1
```

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

**Comment by ShE3py** [comment]
Running the program with the MIR interpreter detects that:
```
$ cargo +nightly miri run
error: the evaluated program deadlocked
  --> /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/pal/unix/futex.rs:80:21
   |
80 |                     )
   |                     ^ this thread got stuck here
   |
   = note: BACKTRACE:
   = note: inside `std::sys::pal::unix::futex::futex_wait` at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/pal/unix/futex.rs:80:21: 80:22
   = note: inside `std::sys::sync::mutex::futex::Mutex::lock_contended` at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/sync/mutex/futex.rs:61:13: 61:53
   = note: inside `std::sys::sync::mutex::futex::Mutex::lock` at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/sync/mutex/futex.rs:33:13: 33:34
   = note: inside `std::sync::Mutex::<i32>::lock` at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sync/poison/mutex.rs:437:13: 437:30
note: inside `main`
  --> src/main.rs:5:44
   |
5  |     println!("{}, {} ", x.lock().unwrap(), x.lock().unwrap());
   |                                            ^^^^^^^^
```
While detecting deadlocks at compile time would be best, I don't think that Rust currently have that.

**Comment by HuMtk** [comment]
> Running the program with the MIR interpreter detects that:使用MIR解释器运行程序检测到：
> 
> ```
> $ cargo +nightly miri run
> error: the evaluated program deadlocked
>   --> /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/pal/unix/futex.rs:80:21
>    |
> 80 |                     )
>    |                     ^ this thread got stuck here
>    |
>    = note: BACKTRACE:
>    = note: inside `std::sys::pal::unix::futex::futex_wait` at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/pal/unix/futex.rs:80:21: 80:22
>    = note: inside `std::sys::sync::mutex::futex::Mutex::lock_contended` at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/sync/mutex/futex.rs:61:13: 61:53
>    = note: inside `std::sys::sync::mutex::futex::Mutex::lock` at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/sync/mutex/futex.rs:33:13: 33:34
>    = note: inside `std::sync::Mutex::<i32>::lock` at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sync/poison/mutex.rs:437:13: 437:30
> note: inside `main`
>   --> src/main.rs:5:44
>    |
> 5  |     println!("{}, {} ", x.lock().unwrap(), x.lock().unwrap());
>    |                                            ^^^^^^^^
> ```
> 
> While detecting deadlocks at compile time would be best, I don't think that Rust currently have that.虽然在编译时检测死锁是最好的，但我认为Rust目前还没有这个功能。

Thanks for your reply. I tried the following. As you said, Rust does not have compile-time deadlock detection. Detecting deadlocks can be a complicated matter. After all, deadlocks always involve runtime.
